auth: During update of auth.allow fuse mount is getting disconnect

mohit84 opened the pull request:
A user has configured huge number of auth.allow addresses(more
than 100) and while a client is trying to mount/remount the volume
the setvolume request has not been finished within threshold
limit (42s).The setvolume is not finished because dns-server is taking
time to resolve the address.The current function (compare_addr_and_update)
logic is it is trying to resolve dns address one by one as configured hosts
in auth-list and then compare with client address.If a client hostname is
configured in the beginning of the auth-list then it will be resolve fast
otherwise it can breach the timeout threshold limit.

Solution: Instead of resolving every auth-address just resolve the peer
          address and compare the peer_addr_hostname with user configured
          auth address limit.In this approach it will call only dns address
          call to compare the address.

Change-Id: I7d59ec6a2c8c3e75e717c8f7ad4e73cfe6582989
Fixes: #2662
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mykaul commented on xlators/protocol/auth/addr/src/addr.c:
```diff
@@ -85,6 +94,7 @@ compare_addr_and_update(char *option_str, char *peer_addr, char *subvol,
 
     ret = -1;
 out:
+    GF_FREE(hostname);
```
```suggestion
    if (hostname)
        GF_FREE(hostname);
```

mohit84 replied:
I don;t think it is required, GF_FREE already check if ptr is NULL or not.

mykaul commented on xlators/protocol/auth/addr/src/addr.c:
```diff
@@ -45,6 +45,16 @@ compare_addr_and_update(char *option_str, char *peer_addr, char *subvol,
+    if (ret) {
+        gf_log(subvol, GF_LOG_ERROR, "DNS ip to host %s resolution is failed",
+               peer_addr);
+        GF_FREE(hostname);
```
Is this needed, if we clean it up in 'out' ?

mohit84 replied:
Not required

mohit84 commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/basic/fencing/afr-lock-heal-advanced.t

2 test(s) needed retry 
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/afr/split-brain-heal-info.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/1468/

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/protocol/bug-1433815-auth-allow.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/basic/afr/granular-esh/granular-esh.t
./tests/bugs/protocol/bug-1433815-auth-allow.t
https://build.gluster.org/job/gh_centos7-regression/1470/

mohit84 commented:
/run regression

xhernandez reviewed (commented):
The patch itself seems fine, but I'm not sure if it will work. Direct and reverse DNS zones are independent, and resolving a name to an IP and the IP to the name can return different things. We can even have multiple names for the same IP and CNAME records that add more complexity.

For example we could have a machine name in auth.allow that corresponds to the machine that is connecting, but the resolution from IP to name could return a different name, and the connection won't be allowed.

A very quick example:

```
# dig www.gluster.org
[...]
;; ANSWER SECTION:
www.gluster.org.        41      IN      CNAME   gluster.wpengine.com.
gluster.wpengine.com.   41      IN      A       35.196.109.67
[...]

# dig -x 35.196.109.67
[...]
;; ANSWER SECTION:
67.109.196.35.in-addr.arpa. 120 IN      PTR     67.109.196.35.bc.googleusercontent.com.
[...]
```

mohit84 committed:
auth: During graph reconfigure a client is getting disconnect

A user has configured huge number of auth.allow addresses(more
than 100) and during graph reconfigure server xlator automatically
validates all the connected client again and in case if a user has
configured huge addresses along with too many clients a brick process
throw too many dns call's to resolve the same address again and again.Because
a server is busy to authenticate connected requests at the same time
client got disconnect if a client is not getting any response within
threshold limit.

Solution: Use dnscache to save the addrinfo of a host(ip/fqdn).
Change-Id: I7d59ec6a2c8c3e75e717c8f7ad4e73cfe6582989
Fixes: #2662
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
> The patch itself seems fine, but I'm not sure if it will work. Direct and reverse DNS zones are independent, and resolving a name to an IP and the IP to the name can return different things. We can even have multiple names for the same IP and CNAME records that add more complexity.
> 
> For example we could have a machine name in auth.allow that corresponds to the machine that is connecting, but the resolution from IP to name could return a different name, and the connection won't be allowed.
> 
> A very quick example:
> 
> ```
> # dig www.gluster.org
> [...]
> ;; ANSWER SECTION:
> www.gluster.org.        41      IN      CNAME   gluster.wpengine.com.
> gluster.wpengine.com.   41      IN      A       35.196.109.67
> [...]
> 
> # dig -x 35.196.109.67
> [...]
> ;; ANSWER SECTION:
> 67.109.196.35.in-addr.arpa. 120 IN      PTR     67.109.196.35.bc.googleusercontent.com.
> [...]
> ```

I have implemented a new approach, use cache to save the dns server to avoid getaddrinfo call again and again,

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+ */
+void
+gf_dns_addr_cache_deinit(struct dnscache *cache)
+{
```
This is using internal dict_t structure. This should never be accessed from outside `libglusterfs/src/dict.c`.

mohit84 replied:
done

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+    struct addrinfo hints;
+    int gai_err = 0;
+
+    if (!dnscache)
```
Can happen ?

xhernandez replied:
???

mohit84 replied:
done

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+
+    cache = dnscache->cache_dict;
+
+    /* Quick cache lookup to see if we already hold it */
```
Currently `dict_t` is implemented as a linked list, so lookups won't be so fast if the cache is big. For small caches, this should be ok.

mohit84 replied:
I think auth_address wont be too long so i believe we can use dict_t.

xhernandez replied:
I think we are also solving client addresses ? do we have some limit on the number of clients we may have ?

Anyway, for now this should be ok.

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+        dnsentry = (struct dnscache_addr_entry *)entrydata->data;
+        /* First check the TTL & timestamp */
+        if (gf_time() - dnsentry->timestamp > dnscache->ttl) {
+            gf_dnscache_addr_entry_deinit(dnsentry);
+            entrydata->data = NULL; /* Mark this as 'null' so
+                                     * dict_del () doesn't try free
+                                     * this after we've already
+                                     * freed it.
+                                     */
+
+            dict_del(cache, identifier); /* Remove this entry */
```
Too hacky IMO, and it's accessing internal `dict_t` structure.

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+
+    /* Get the addr */
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
```
This can return both IPv4 and IPv6 addresses mixed. Are all the callers ready to handle that correctly ?

mohit84 replied:
Yes currently we are using it only two places, the one place for to get hostname and 2nd for to compare two address so it will work fine.

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+    if (addr && !from_cache && identifier) {
+        struct dnscache_addr_entry *entry = gf_dnscache_addr_entry_init();
+
+        if (entry) {
```
It's fine to not return an error if the `entry` creation fails because we already have the needed information. But in this case, who will release the memory allocated by for `addr` ?

mohit84 replied:
done

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+        if (entry) {
+            entry->addr = addr;
+            entry->timestamp = gf_time();
+            entrydata = bin_to_data(entry, sizeof(*entry));
+            dict_set(cache, identifier, entrydata);
```
`bin_to_data()` and `dict_set()` could fail.

mohit84 replied:
DONE

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -3686,18 +3804,26 @@ gf_is_same_address(char *name1, char *name2)
-                gai_strerror(gai_err), NULL);
-        goto out;
+    if (!p1) {
+        gai_err = getaddrinfo(name1, NULL, &hints, &addr1);
```
Why not using `gf_dns_lookup_address_cached()` to get it from cache (or add it if not already present) ?

mohit84 replied:
Because gf_is_same_address can be call by non brick process also currently dnscache has been populated by some xlator either server or iostat.

xhernandez replied:
And what's the problem ? do we need to resolve to different addresses based on if the caller is a brick or not ?

mohit84 replied:
The problem is currently dnscache is declared only for server/io-stat xlator, in case of glusterd it won't be able to access the dnscache.We can access dnscache in this function only if it has declared on glusterfs_ctx_t.If i do declare dnscache on glusterfs_ctx_t the other issue is dnscache_ttl_sec, currently the option is configurable for server and io-stat xlator only.

mohit84 replied:
oh i missed it, you wanted to check if dnscache is valid on xlator ptr then we can use cache otherwise need to call dns call.
I shall update the same in next patch.

xhernandez replied:
> The problem is currently dnscache is declared only for server/io-stat xlator, in case of glusterd it won't be able to access the dnscache.We can access dnscache in this function only if it has declared on glusterfs_ctx_t.

Yes, that's what I think we should do. The DNS cache is a global thing, not per-xlator.

> If i do declare dnscache on glusterfs_ctx_t the other issue is dnscache_ttl_sec, currently the option is configurable for server and io-stat xlator only.

The option should be global. We already have some global options, but I'm not sure if they are shared with glusterd. If not, we could read the option from glusterd.vol for glusterd.

> oh i missed it, you wanted to check if dnscache is valid on xlator ptr then we can use cache otherwise need to call dns call.

No, I don't think this is useful. The cache should be global. We don't want to resolve the same name once for each xlator that requests it.

mohit84 replied:
I have declared dnscache as a part of glusterfs_ctx, for the time being i am using only for brick process.
Later we can use same cache for rpc_trans and glusterd also.

xhernandez commented on xlators/protocol/auth/addr/src/addr.c:
```diff
@@ -59,7 +66,9 @@ compare_addr_and_update(char *option_str, char *peer_addr, char *subvol,
         length = strlen(addr_str);
         if ((addr_str[0] != '*') && valid_host_name(addr_str, length)) {
-            match = gf_is_same_address(addr_str, peer_addr);
+            addr2 = NULL;
```
???

mohit84 replied:
done

xhernandez reviewed (changes requested):
Why not modify the existing cache to handle `struct addrinfo` data instead of a single `ip` and `fqdn` ? it seems like we are duplicating unnecessary things.

Additionally, if the dns cache functions are defined in common-utils, why the cache itself is not made global so that all xlators/components can make use of it ? is there any benefit in having multiple caches ?

mohit84 committed:
auth: Resolved several reviewer comments

Fixes: #2662
Change-Id: I008e59d1b335185c9487b997640a906e50e76a75
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 committed:
auth: Resolved several reviewer comments

Fixes: #2662
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: I5f4c22bfccc109aefc369c00809d8d6d0a911fc8

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

xhernandez commented on glusterfsd/src/glusterfsd.c:
```diff
@@ -1739,6 +1745,7 @@ glusterfs_ctx_defaults_init(glusterfs_ctx_t *ctx)
         mem_pool_destroy(ctx->dict_data_pool);
         mem_pool_destroy(ctx->dict_pair_pool);
         mem_pool_destroy(ctx->logbuf_pool);
+        gf_dnscache_deinit(ctx->dnscache);
```
`ctx->dnscache` will always be `NULL` here.

mohit84 replied:
done

xhernandez commented on glusterfsd/src/glusterfsd.c:
```diff
@@ -2666,10 +2673,8 @@ GF_IO_ASYNC(main_start, op, static)
-        .setup = main_start,
-        .cleanup = main_terminate
-    };
+    gf_io_handlers_t main_handlers = {.setup = main_start,
+                                      .cleanup = main_terminate};
```
Please, don't change this. I know that clang-format wants to change it, but it becomes less readable IMO.

mohit84 replied:
The change was not done by me, it was done by clang.

mohit84 replied:
done

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -3682,22 +3699,38 @@ gf_is_same_address(char *name1, char *name2)
+    gf_boolean_t flag = _gf_false;
+
+    if (ctx && ctx->dnscache)
+        flag = _gf_true;
```
It would be better to use a more meaningful name, like "have_cache" or something like that.

mohit84 replied:
done

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -3682,22 +3699,38 @@ gf_is_same_address(char *name1, char *name2)
+        if (!addr2)
+            return ret;
+    } else {
+        gai_err = getaddrinfo(name1, NULL, &hints, &addr1);
+        if (gai_err != 0) {
+            gf_smsg(name1, GF_LOG_WARNING, 0, LG_MSG_GETADDRINFO_FAILED,
+                    "error=%s", gai_strerror(gai_err), NULL);
+            goto out;
+        }
+
+        gai_err = getaddrinfo(name2, NULL, &hints, &addr2);
+        if (gai_err != 0) {
+            gf_smsg(name2, GF_LOG_WARNING, 0, LG_MSG_GETADDRINFO_FAILED,
+                    "error=%s", gai_strerror(gai_err), NULL);
+            goto out;
+        }
```
This code is duplicated, and it's the same used in `gf_dns_lookup_address_cached()`. It would be better to move it to a new function and reuse it.

mohit84 replied:
done

xhernandez commented on xlators/debug/io-stats/src/io-stats.c:
```diff
@@ -1080,9 +1077,10 @@ _io_stats_write_latency_sample(xlator_t *this, ios_sample_t *sample,
-            hostname = "Unknown";
+        addr = gf_dns_lookup_address_cached(identifier, ctx->dnscache);
+        if (addr)
+            getnameinfo(addr->ai_addr, addr->ai_addrlen, hostname,
```
Why not use `addr->ai_canonname` instead of another request ? (`AI_CANONNAME` flag will be needed in getaddrinfo).

mohit84 replied:
ai_canonname does not work in case of ip_address, it works only while identifier is a hostname.

xhernandez replied:
In that case, why we do need to call `gf_dns_lookup_address_cached()` ? if `identifier` is already a hostname, we don't need to do anything. If it's an IP, we can directly call `getnameinfo()`. Probably we should incorporate `getnameinfo()` into the cached information so that we can resolve from host to IP and from IP to host.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volume-set.c:
```diff
@@ -1684,7 +1684,11 @@ struct volopt_map_entry glusterd_volopt_map[] = {
         .value = "1024",
     },
-
+    {
+        .key = "server.dnscache-ttl-sec",
+        .voltype = "protocol/server",
+        .op_version = GD_OP_VERSION_10_0,
+    },
```
Is this used ?

mohit84 replied:
No

xhernandez commented on xlators/debug/io-stats/src/io-stats.c:
```diff
@@ -3828,16 +3826,18 @@ mem_acct_init(xlator_t *this)
 
 void
-ios_conf_destroy(struct ios_conf *conf)
+ios_conf_destroy(xlator_t *this)
```
Any reason for changing this ?

mohit84 replied:
done

mohit84 committed:
auth: Resolved several reviewer comments

Fixes: #2662
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: I78668e2e72f18589eae8b25c8fb04cd1181d8db2

mohit84 commented:
/run regression

mohit84 commented:
/run regression

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -637,82 +637,111 @@ gf_dnscache_entry_init()
+        return addr;
+
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
```
After reading the manual, I think we should use `AI_ADDRCONFIG` flag in `hints.ai_flags`

gluster-ant commented:
1 test(s) failed 
./tests/basic/fop-sampling.t

0 test(s) generated core 


5 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/afr/split-brain-healing-ctime.t
./tests/basic/ec/ec-reset-brick.t
./tests/basic/fop-sampling.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/1530/

mohit84 committed:
auth: Resolved several reviewer comments

Fixes: #2662
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

Change-Id: Ia2302e8ea759381a21ad5e43b0c38fedbce98377

mohit84 commented:
/run regression

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/basic/fop-sampling.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/basic/afr/ta-shd.t
./tests/basic/fop-sampling.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/1533/

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/basic/nufa.t

1 test(s) generated core 
./tests/basic/nufa.t

2 test(s) needed retry 
./tests/basic/ctime/ctime-ec-heal.t
./tests/basic/nufa.t
https://build.gluster.org/job/gh_centos7-regression/1534/

mohit84 commented:
/run regression

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index 08b81b010..20e5ea2ed 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -759,7 +759,7 @@ gf_getaddrinfo(char *identifier)
     } else {
         memset(&hints, 0, sizeof(hints));
         hints.ai_family = AF_UNSPEC;
-        hints.ai_flags = (AI_CANONNAME|AI_ADDRCONFIG);
+        hints.ai_flags = (AI_CANONNAME | AI_ADDRCONFIG);
 
         ret = getaddrinfo(identifier, NULL, &hints, &addr);
         if (ret != 0) {
```

gluster-ant commented:
1 test(s) failed 
./tests/bugs/cli/bug-1169302.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/basic/afr/ta-shd.t
./tests/basic/glusterd/heald.t
./tests/bugs/bitrot/bug-1227996.t
./tests/bugs/cli/bug-1169302.t
https://build.gluster.org/job/gh_centos7-regression/1535/

mohit84 commented:
/run full regression

gluster-ant commented:
3 test(s) failed 
./tests/bugs/cli/bug-1169302.t
./tests/bugs/protocol/bug-1321578.t
./tests/bugs/protocol/bug-1433815-auth-allow.t

0 test(s) generated core 


8 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/afr/ta-shd.t
./tests/basic/volume-snapshot-clone.t
./tests/bugs/cli/bug-1169302.t
./tests/bugs/protocol/bug-1321578.t
./tests/bugs/protocol/bug-1433815-auth-allow.t

3 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/glusterd-restart-shd-mux.t

xhernandez commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/cli/bug-1169302.t

0 test(s) generated core 


1 test(s) needed retry 
./tests/bugs/cli/bug-1169302.t
https://build.gluster.org/job/gh_centos7-regression/1577/

xhernandez commented:
/run regression

mohit84 commented:
/run regression

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index f3d61b5d1..a7258b930 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -759,7 +759,7 @@ gf_getaddrinfo(char *identifier)
     } else {
         memset(&hints, 0, sizeof(hints));
         hints.ai_family = AF_UNSPEC;
-        hints.ai_flags = (AI_CANONNAME|AI_ADDRCONFIG);
+        hints.ai_flags = (AI_CANONNAME | AI_ADDRCONFIG);
 
         ret = getaddrinfo(identifier, NULL, &hints, &addr);
         if (ret != 0) {
```

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/cli/bug-1169302.t

0 test(s) generated core 


3 test(s) needed retry 
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/afr/ta-shd.t
./tests/bugs/cli/bug-1169302.t
https://build.gluster.org/job/gh_centos7-regression/1601/
